pkg/query/caveat.go

```diff
@@ -0,0 +1,218 @@
+package query
+
+import (
+	"fmt"
+
+	core "github.com/authzed/spicedb/pkg/proto/core/v1"
+)
+
+// CaveatIterator wraps another iterator and applies caveat evaluation to its results.
+// It checks caveat conditions on relationships during iteration and only yields
+// relationships that satisfy the caveat constraints.
+type CaveatIterator struct {
+	subiterator Iterator
+	caveat      *core.ContextualizedCaveat
+}
+
+var _ Iterator = &CaveatIterator{}
+
+// NewCaveatIterator creates a new caveat iterator that wraps the given subiterator
+// and applies the specified caveat conditions.
+func NewCaveatIterator(subiterator Iterator, caveat *core.ContextualizedCaveat) *CaveatIterator {
+	return &CaveatIterator{
+		subiterator: subiterator,
+		caveat:      caveat,
+	}
+}
+
+func (c *CaveatIterator) CheckImpl(ctx *Context, resources []Object, subject ObjectAndRelation) (PathSeq, error) {
+	subSeq, err := ctx.Check(c.subiterator, resources, subject)
+	if err != nil {
+		return nil, err
+	}
+	return c.runCaveats(ctx, subSeq)
+}
+
+func (c *CaveatIterator) IterSubjectsImpl(ctx *Context, resource Object) (PathSeq, error) {
+	subSeq, err := ctx.IterSubjects(c.subiterator, resource)
+	if err != nil {
+		return nil, err
+	}
+	return c.runCaveats(ctx, subSeq)
+}
+
+func (c *CaveatIterator) IterResourcesImpl(ctx *Context, subject ObjectAndRelation) (PathSeq, error) {
+	subSeq, err := ctx.IterResources(c.subiterator, subject)
+	if err != nil {
+		return nil, err
+	}
+
+	return c.runCaveats(ctx, subSeq)
+}
+
+func (c *CaveatIterator) runCaveats(ctx *Context, subSeq PathSeq) (PathSeq, error) {
+	if c.caveat == nil {
+		return subSeq, nil
+	}
+
+	return func(yield func(Path, error) bool) {
+		ctx.TraceStep(c, "applying caveat '%s' to sub-iterator results", c.caveat.CaveatName)
+
+		processedCount := 0
+		passedCount := 0
+
+		for path, err := range subSeq {
+			if err != nil {
+				if !yield(path, err) {
+					return
+				}
+				continue
```

Comment:
I am confused by the `continue` here and in the error case below - should it not be `return` instead?